### PR TITLE
Logging: log HTTP requests and replies

### PR DESCRIPTION
**Pull request by reinkrul**
Closes https://github.com/nuts-foundation/nuts-node/issues/1177

Introduces `Log` config for HTTP endpoints which can be configured to log only metadata (does it already do, default setting), log nothing or log metadata and request-reply.

Example output:

```
INFO[0004] HTTP request                                  method=POST module=http-server remote_ip="::1" status=400 uri=/n2n/auth/v1/accesstoken
INFO[0004] HTTP request body: grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Ajwt-bearer&assertion=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJ1cm46b2lkOjIuMTYuODQwLjEuMTEzODgzLjIuNC42LjE6NDgwMDAwMDAiLCJzdWIiOiJ1cm46b2lkOjIuMTYuODQwLjEuMTEzODgzLjIuNC42LjE6MTI0ODEyNDgiLCJzaWQiOiJ1cm46b2lkOjIuMTYuODQwLjEuMTEzODgzLjIuNC42LjM6OTk5OTk5MCIsImF1ZCI6Imh0dHBzOi8vdGFyZ2V0X3Rva2VuX2VuZHBvaW50IiwidXNpIjoiYmFzZTY0IGVuY29kZWQgc2lnbmF0dXJlIiwiZXhwIjoxNTc4MTEwNDgxLCJpYXQiOjE1Nzg5MTA0ODEsImp0aSI6IjEyMy00NTYtNzg5In0.76XtU81IyR3Ak_2fgrYsuLcvxndf0eedT1mFPa-rPXk  module=http-server
INFO[0004] HTTP response body: {"error":"invalid_request","error_description":"jwt bearer token validation failed: invalid key ID (id=): invalid DID: input length is less than 7"}  module=http-server
```

If solution is OK, I'll write some unit tests.

### CODE REVIEW

**Review by gerardsn — COMMENTED** [review]
From a debugging perspective I think this is fine but it is probably too verbose to use in production.
If the increased verbosity is mainly for debugging I would change the flag names to reflect this.

**Comment by reinkrul** [comment]
> From a debugging perspective I think this is fine but it is probably too verbose to use in production. If the increased verbosity is mainly for debugging I would change the flag names to reflect this.

Or analysis on acceptation/production what is going wrong. But it might log PII, so you need to take great care enabling this on environments with real data. I don't think we should mention that, I expect people to be configuring healthcare production environments to think about that before fiddling with config?

File: http/requestlogger.go [hunk 1]
```diff
@@ -60,13 +50,53 @@
...
+				requestBody = string(request)
+			}
+
+			responseContentType := e.Response().Header().Get("Content-Type")
+			responseBody := "(not loggable: " + responseContentType + ")"
+			if isLoggableContentType(responseContentType) {
+				responseBody = string(response)
+			}
+
+			logger.Infof("HTTP request body: %s", requestBody)
```
[the remark this answers is not in the document]
Comment (woutslakhorst) [reply http/requestlogger.go:82]: CodeQL is correct here, the body should be sanitized.

File: http/engine.go [hunk 2]
```diff
@@ -277,6 +272,23 @@ func (h Engine) applyBindMiddleware(echoServer EchoServer, path string, excludeP
...
+		return matchesPath(c.Request().RequestURI, "/metrics") || matchesPath(c.Request().RequestURI, "/status")
+	}
+	echoServer.Use(requestLoggerMiddleware(func(c echo.Context) bool {
+		// Log when level is set to metadata or request-reply
+		return cfg.Log == LogNothingLevel || loggerSkipper(c)
+	}, log.Logger()))
+	echoServer.Use(bodyLoggerMiddleware(func(c echo.Context) bool {
+		// Log when level is set to request-reply
+		return cfg.Log != LogRequestReplyLevel || skipper(c)
+	}, log.Logger()))
```
Comment (gerardsn) [inline http/engine.go:290]: when `cfg.Log == LogNothingLevel`, no request will ever be logged
when `cfg.Log != LogRequestReplyLevel`, no body will ever be logged

Do we even need to register the middleware in these cases?

Comment (reinkrul) [reply http/engine.go:290]: Probably not.

**Review by gerardsn — CHANGES_REQUESTED** [review]

File: http/config.go [hunk 3]
```diff
@@ -48,8 +49,22 @@ type InterfaceConfig struct {
...
+// LogLevel specifies what to log for incoming/outgoing HTTP traffic.
+type LogLevel string
+
+const (
+	// LogNothingLevel indicates nothing will be logged for incoming/outgoing HTTP traffic.
+	LogNothingLevel LogLevel = "nothing"
+	// LogMetadataLevel indicates that only metadata (HTTP URI, method, response code, etc) will be logged for incoming/outgoing HTTP traffic.
+	LogMetadataLevel = "metadata"
+	// LogRequestReplyLevel indicates that metadata and full request/reply bodies  will be logged for incoming/outgoing HTTP traffic.
+	LogRequestReplyLevel = "request-reply"
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply http/config.go]: Changed it to `metadata-and-body`

**Review by gerardsn — APPROVED** [review]

**Pull request merged**